subsys/net/lib/dns/mdns_responder.c

```diff
@@ -606,6 +650,442 @@ static int dns_read(int sock,
+		mon_if[first_free].updated = true;
+		mon_if[first_free].iface = iface;
+
+		memcpy(&mon_if[first_free].addr.in_addr, address, expected_len);
+
+		NET_DBG("%s added %s address %s to iface %d in slot %d",
+			"mDNS", family == AF_INET ? "IPv4" : "IPv6",
+			family == AF_INET ? net_sprint_ipv4_addr(address) :
+					    net_sprint_ipv6_addr(address),
+			net_if_get_by_iface(iface), first_free);
+		return 0;
+	}
+
+	return -ENOENT;
+}
+
+static int del_address(struct net_if *iface, sa_family_t family,
+		       const void *address, size_t addrlen)
+{
+	size_t expected_len;
+
+	if (family != AF_INET && family != AF_INET6) {
+		return -EINVAL;
+	}
+
+	if (family == AF_INET) {
+		expected_len = sizeof(struct in_addr);
+	} else {
+		expected_len = sizeof(struct in6_addr);
+	}
+
+	ARRAY_FOR_EACH(mon_if, j) {
+		if (!mon_if[j].in_use) {
+			continue;
+		}
+
+		if (mon_if[j].iface != iface) {
+			continue;
+		}
+
+		if (mon_if[j].addr.family != family) {
+			continue;
+		}
+
+		if (memcmp(&mon_if[j].addr.in_addr, address, expected_len) == 0) {
+			mon_if[j].added = false;
+			mon_if[j].updated = true;
+			return 0;
+		}
+	}
+
+	return -ENOENT;
+}
+
+static void probe_cb(enum dns_resolve_status status,
+		     struct dns_addrinfo *info,
+		     void *user_data)
+{
+	struct mdns_probe_user_data *ud = user_data;
+
+	NET_DBG("status %d", status);
+
+	if (status == DNS_EAI_ALLDONE) {
+		/* We got a reply, so we can stop probing as there was reply
+		 * to our query.
+		 */
+		cancel_probes(ud->ctx);
+		failed_probes++;
+
+		NET_WARN("mDNS probe received data, will not use \"%s\" name.",
+			 ud->query);
+
+	} else if (status == DNS_EAI_CANCELED) {
+		if (--probe_count <= 0) {
+			int ret;
+
+			/* Schedule a handler that will initialize the listener
+			 * and start announcing the services.
+			 */
+			ret = k_work_reschedule_for_queue(&mdns_work_q,
+							  &init_listener_timer,
+							  K_NO_WAIT);
+			if (ret < 0) {
+				NET_DBG("Cannot schedule %s init work (%d)", "mDNS", ret);
+			}
+		}
+	}
+
+	dns_resolve_close(&ud->ctx->probe_ctx);
+}
+
+#define PROBE_TIMEOUT 1750 /* in ms, see RFC 6762 ch 8.1 */
+#define PORT_COUNT 5
+
+/* Sending unsolicited mDNS query when interface is up. RFC 6762 ch 8.1 */
+static int send_probe(struct mdns_responder_context *ctx)
+{
+	const char *hostname = net_hostname_get();
+	struct sockaddr *servers[2] = { 0 };
+	struct sockaddr_in6 server = { 0 };
+	int interfaces[ARRAY_SIZE(servers)] = { 0 };
+	uint16_t local_port;
+	int ret;
+
+	NET_DBG("%s %s %s to our hostname %s.local iface %d", "mDNS",
+		ctx->dispatcher.local_addr.sa_family == AF_INET ? "IPv4" : "IPv6",
+		"probe", hostname, net_if_get_by_iface(ctx->iface));
+
+	ret = 0;
+	do {
+		local_port = sys_rand16_get() | 0x8000;
+		ret++;
+	} while (net_context_port_in_use(IPPROTO_UDP, local_port,
+					 &ctx->dispatcher.local_addr) && ret < PORT_COUNT);
+	if (ret >= PORT_COUNT) {
+		NET_ERR("No available port, %s probe fails!", "mDNS");
+		ret = -EIO;
```

Comment:
We're mixing errno and DNS error space here, in error checks below we set `DNS_EAI_SYSTEM` as a return value. Shouldn't we set errno and return DNS_EAI_SYSTEM here too?

Reply:
My bad, I mixed DNS and errno space. We should just use the returned value directly, no need to do DNS_EAI_SYSTEM here.